rm-datasource/src/main/java/io/seata/rm/datasource/sql/SQLRecognizerGroupFactory.java

```diff
@@ -0,0 +1,55 @@
+/*
+ *  Copyright 1999-2019 Seata.io Group.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package io.seata.rm.datasource.sql;
+
+import io.seata.common.loader.EnhancedServiceLoader;
+import java.text.MessageFormat;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+/**
+ * The SQLRecognizerGroupFactory
+ *
+ * @author: Zhibei Hao丶
+ * @date: 2019/11/8 17:37
+ * @version: V1.0
+ */
+public class SQLRecognizerGroupFactory
+{
+
+  private static Object lockObj = new Object();
+  private static Map<String, SQLRecognizerGroup> recognizerGroupMap;
+
+  public static SQLRecognizerGroup getSQLRecognizerGroup(String dbType) {
+
+    if (recognizerGroupMap == null) {
```

Comment:
It is recommended to optimize the concurrency when returning an empty map or the groupList not put completely.

Reply:
OK, I've made a note. Please continue to review other points. I will make unified changes tomorrow, because my local code is on another company's computer.

Reply:
Hello, I reviewed the code and  want to explian it. This part of code will only be executed once after the application is started: when obtaining SQLRecognizer for the first time, if there is concurrency (such as executing SQL at multiple places at the same time), the thread obtained the lock will do the following:
1. new HashMap.
2. Load All  SQLRecognizerGroup implementations and put them all into the groupMap at once.
Therefore, groupmap has only two states: it is empty or it already contains all the implementation of SQLRecognizerGroup. There is no intermediate state.

Reply:
e.g.
一
1. thread1 getLock, recognizerGroupMap = new HashMap<>();
2.thread2 recognizerGroupMap!=null but is empty
二
1.thread1 getLock, recognizerGroupMap.put not completely
2 thread2 recognizerGroupMap.containsKey maybe false

Reply:
Thank you for your explanation. It was really my negligence that led to this bug
Now i modify the code as follows:

//initialized is set to false
if (!initialized) {
  synchronized (SQLRecognizerGroupFactory.class) {
    if (!initialized) {
      1.new groupMap
      2.put values into map
      3.set initialized=true
    }
  }
}
if (recognizerGroupMap.containsKey(dbType)) {
  return recognizerGroupMap.get(dbType);
}
throw ex

And The code has been synchronized to my PR, please review,Thanks

Reply:
I further optimized the code:removed the initialized variable, and delayed the assignment of groupmap until the completion of put.